### DISCUSSION TITLE
Instructions Needed for Using Dockerfile to Build a UI-Compatible Engine

### DISCUSSION THREAD

**Discussion by Xanonymous-GitHub** [dhbloo/rapfi#6] [Q&A] [answered]
Dear @dhbloo,

Firstly, I want to express my appreciation for your work on the Rapfi engine. It is a very impressive project, and I am excited to use it.

Currently, I am trying to create a Docker environment to build and run Rapfi. I have constructed a Dockerfile that compiles and runs the engine successfully within the container. Here is the Dockerfile for your reference:

```dockerfile
# Builder Stage
FROM alpine:edge as builder

# Install necessary packages using apk add
RUN apk update && apk add \
    cmake \
    g++ \
    curl \
    unzip \
    git \
    libtbb-dev \
    make \
    linux-headers

# Working directory setup
WORKDIR /app

# Clone the repository
RUN git clone https://github.com/dhbloo/rapfi.git . \
    && git submodule update --init --recursive --force && git submodule sync

# Build process
RUN cd Rapfi \
    && mkdir build  \
    && cd build  \
    && cmake .. -DCMAKE_BUILD_TYPE=Release \
    && make

# Final Stage
FROM alpine:edge
LABEL author="xanonymous"

# Copy necessary files from builder stage
COPY --from=builder /app/Rapfi/build/pbrain-rapfi /app/rapfi
COPY --from=builder /app/Networks/config-example/config.toml /app/config.toml
COPY --from=builder /app/Networks/mix7nnue /app
COPY --from=builder /app/Networks/classical /app

# Install necessary packages using apk add
RUN apk update && apk add --no-cache \
    libtbb-dev \
    linux-headers \
    && sed -i 's/default_thread_num = 1/default_thread_num = 0/g' /app/config.toml

# Set default command to execute
CMD ["/app/rapfi"]
```

I am now trying to figure out how to make this Docker-built Rapfi engine communicate with a UI, so that I can use it to play games. However, I am unsure of the necessary steps to make this happen.

Could you please provide some guidance on how I could go about doing this? Specifically, I need to understand how the UI should interact with the Rapfi engine running within the Docker container.

Any guidance or advice you could provide would be greatly appreciated. Thank you for your time and consideration.

**Comment by dhbloo** [comment]
Hello,
As the UI communicates with Rapfi using the Piskvork protocol through standard input&output, it seems you need to set up a pipe to send commands into STDIN and get results from STDOUT of the container. Also, you need to build a UI that supports using pipes from the docker container as its engine backend.

I am not familiar with docker, but I searched something like [this](https://stackoverflow.com/questions/59833276/write-to-stdin-of-running-docker-container) to send something into the stdin of a container, FYI.

**Reply by Xanonymous-GitHub** [reply]
Yes, Docker allows capturing the standard input (STDIN) and output (STDOUT) streams from a running container, which would indeed facilitate communication with the Rapfi engine. However, my primary challenge now is understanding the specifics of the communication protocol - the Piskvork protocol - that the UI should use to communicate with the engine.

I tried accessing the link you provided for the Piskvork protocol, but unfortunately, it seems to be broken or inaccessible, showing an error when accessed.

Would it be possible for you to provide more details about the Piskvork protocol directly, or perhaps direct me to an alternate resource where I could learn more about it? Understanding the mechanics of this protocol is crucial for me to correctly configure the UI and ensure seamless communication with the engine.

BTW, here is the built docker image, you may take it if you need:
[xanonymous/rapfi](https://hub.docker.com/r/xanonymous/rapfi/)
<img width="574" alt="截圖 2023-07-25 上午9 47 18" src="https://github.com/dhbloo/rapfi/assets/47718989/50238776-21b1-47a2-8683-8f7c36b4b62f">

**Reply by dhbloo** [reply, marked as answer]
It seems the page has been moved to [a new address](https://plastovicka.github.io/protocl2en.htm). In addition to this, [Yixin board extension](https://github.com/accreator/Yixin-protocol) describes additional interaction about multi-threading, etc. There are also some commands unique to rapfi, which I will add a detailed description in readme soon.

The docker image is nice. I think I will add more detailed instructions about running in a docker later.

**Reply by Xanonymous-GitHub** [reply]
Thanks for your help :)
